### PR TITLE
bugfix: mutable state should be reset if the operation is not success…

### PR DESCRIPTION
**Pull request by wxing1292**
…fully

### CODE REVIEW

File: service/history/historyCache_test.go [hunk 1]
```diff
@@ -76,37 +88,72 @@ func (s *historyCacheSuite) SetupTest() {
...
+func (s *historyCacheSuite) TestHistoryCacheClear() {
+	s.mockShard.GetConfig().HistoryCacheMaxSize = 20
+	domainID := "test_domain_id"
+	s.cache = newHistoryCache(s.mockShard, s.logger)
+	we := workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr("wf-cache-test-clear"),
+		RunId:      common.StringPtr(uuid.New()),
+	}
+
+	context, release, err := s.cache.getOrCreateWorkflowExecution(domainID, we)
```
Comment (venkat1109) [inline service/history/historyCache_test.go]: can you spin up 2 or more go-routines that do the same set of things you do below ? that way, if there is any race condition, there is a good chance of finding it at unit test time with race detector

File: service/history/historyCache_test.go [hunk 2]
```diff
@@ -76,37 +88,72 @@ func (s *historyCacheSuite) SetupTest() {
...
 	// Old context should be evicted.
-	newContext, release, err4 := s.cache.getOrCreateWorkflowExecution(domain, we)
+	newContext, release, err4 := s.cache.getOrCreateWorkflowExecution(domainID, we)
 	s.Nil(err4)
 	s.False(context == newContext)
-	release()
+	release(err4)
+}
+
+func (s *historyCacheSuite) TestHistoryCacheClear() {
```
Comment (venkat1109) [inline service/history/historyCache_test.go]: [NOT A BLOCKER for this CR ] can you also add a test that exercises mutable state reload code path (if it doesn't exist already ?) i.e. cache hit followed by clear() due to error followed by mutable state reload

File: service/history/historyCache.go [hunk 3]
```diff
@@ -118,8 +118,11 @@ func (c *historyCache) getOrCreateWorkflowExecution(domainID string,
 	// This will create a closure on every request.
 	// Consider revisiting this if it causes too much GC activity
 	status := cacheNotReleased
-	releaseFunc := func() {
+	releaseFunc := func(err error) {
```
Comment (venkat1109) [inline service/history/historyCache.go]: Like the fact that you changed the signature.. that's the best way to find out all places this is being used at compile time 👍

File: service/history/historyCache_test.go (hunk 1 above)
Comment (wxing1292) [reply service/history/historyCache_test.go]: tested in the update

File: service/history/historyCache_test.go (hunk 2 above)
Comment (wxing1292) [reply service/history/historyCache_test.go]: done in the update

File: service/history/historyCache.go [hunk 4]
```diff
@@ -118,8 +118,11 @@ func (c *historyCache) getOrCreateWorkflowExecution(domainID string,
...
 		if atomic.CompareAndSwapInt32(&status, cacheNotReleased, cacheReleased) {
+			if err != nil {
```
Comment (samarabbas) [inline service/history/historyCache.go]: Let's create a workitem to review if we can skip clearing the cache for certain errors.  Like EntityNotFound.

**Pull request merged**